mitiq/zne/scaling/layer_scaling.py

```diff
@@ -0,0 +1,74 @@
+# Copyright (C) 2023 Unitary Fund
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+"""Functions for layer-wise unitary folding on supported circuits."""
+from typing import List
+from cirq import Circuit, inverse, Moment
+
+
+def layer_folding(circuit: Circuit, layers_to_fold: List[int]) -> Circuit:
+    """Applies a variable amount of folding to select layers of a circuit.
+
+    Args:
+        circuit: The input cirq circuit.
+        layers_to_fold: A list with the index referring to the layer number,
+                        and the element filled by an integer represents the
+                        number of times the layer is folded.
+
+    Returns:
+        A cirq ``Circuit`` with layers and number of times to fold specified
+        by ``layers_to_invert``.
+    """
+    layers = []
+    for i, layer in enumerate(circuit):
+        layers.append(layer)
+        # Apply the requisite number of folds to each layer.
+        num_fold = layers_to_fold[i]
+        for _ in range(num_fold):
+            layers.append(Moment(inverse(layer)))
+            layers.append(Moment(layer))
```

Comment:
Is there a reason we need to call `Moment` in the two append calls, but not at the top? My local testing indicates they might be unnecessary.

Reply:
`mypy` seems to throw a fit if this isn't explicitly stated here. Removing the `Moment()` wrapper and running `make check-types` yields:

```
Argument 1 to "append" of "list" has incompatible type "Union[Operation, OpTree]"; expected "Moment"  [arg-type]
````

We could also explicitly ignore the line as well. Any preference?

Reply:
Weird, is it complaining about appending the layer, or the inverse (or both)? From my testing they should all be moments!

```py
c = cirq.Circuit()
q = cirq.LineQubit.range(2)

c.append(cirq.ops.H.on_each(*q))
c.append(cirq.ops.X.on_each(*q))

for layer in c:
    print(type(layer))
    print(type(cirq.inverse(layer)))

>>> <class 'cirq.circuits.moment.Moment'>
>>> <class 'cirq.circuits.moment.Moment'>
>>> <class 'cirq.circuits.moment.Moment'>
>>> <class 'cirq.circuits.moment.Moment'>
```

What does [`reveal_type(layer)`](https://mypy.readthedocs.io/en/stable/type_inference_and_annotations.html?highlight=reveal_type#type-inference) show?

Reply:
Yeah, it's complaining about both the layer and the inverse. Indeed, when I check the types of the layer and inverse, I see `Moment` as well. I think this might be a quirk of `mypy` or perhaps the version that is pinned is getting confused as without this explicit typecast it's failing in the pipeline.

Reply:
WDYT of this?

```py
if not cirq.is_measurement(layer):
    layers.append(layer**-1)
    layers.append(layer)
```

The advantage of using [`cirq.inverse`](https://github.com/quantumlib/Cirq/blob/v1.1.0/cirq-core/cirq/protocols/inverse_protocol.py#L70-L121) seems to be that it ensures `__pow__` is defined, but I believe this should always be the case for moments that are not measurements, right?